realm/realm-library/src/main/cpp/util.hpp

```diff
@@ -58,11 +61,73 @@ JNIEXPORT jint JNICALL JNI_OnLoad(JavaVM *vm, void *reserved);
 #define STRINGIZE_DETAIL(x) #x
 #define STRINGIZE(x) STRINGIZE_DETAIL(x)
 
-// Exception handling
-#define CATCH_STD() \
-    catch (...) { \
-        ConvertException(env, __FILE__, __LINE__); \
+void ConvertException(JNIEnv* env);
+
+// Exception handling using lambdas
+// Credit: http://stackoverflow.com/questions/29202377/how-do-i-force-the-compiler-to-recognize-lambda-type-in-templated-function
+template<typename T, typename F>
+T try_catch(JNIEnv* env, F func) {
+    try {
+        // if an exception already has occurred, it is not possible to call any JNI
+        // functions, and we don't wish to try to execute the JNI function
+        // alternatively, we could clear the exception but then we can easily miss
+        // an error stage
+        if (env->ExceptionCheck() == JNI_FALSE) {
+            // the actual JNI code is executed
+            // it might throw C++ exceptions
+            return func();
+        }
     }
+    catch (std::exception& e) {
+        // we don't wish to do the exception handling in here; this is a templated function, and we might end up
+        // with multiple copies of the code for no good reason
+        ConvertException(env);
+    }
+
+    // when a Java exception is thrown, we return to Java with any return value; as soon as the execution point is
+    // back at the JVM, the Java exception is thrown, and the returned value isn't used for anything
```

Comment:
Yes,  `get_source_ndx` seems to be the only case. So we don't throw in that situation.